lib/migration_helper.rb

```diff
@@ -51,6 +51,42 @@ def estimate_batch_complete(start_time, progress)
   end
   private :estimate_batch_complete
 
+  #
+  # Views
+  #
+
+  def create_view(name, query, sequence)
+    execute(<<-SQL)
+      CREATE VIEW #{name} AS #{query}
+    SQL
+    execute(<<-SQL)
+      ALTER VIEW #{name} ALTER COLUMN id SET DEFAULT nextval('#{sequence}')
+    SQL
+  end
+
+  def create_metrics_view(name)
+    create_view(name, "SELECT * FROM #{name}_base", "#{name}_base_id_seq")
+  end
+
+  def drop_view(name)
+    execute(<<-SQL)
+      DROP VIEW #{name}
+    SQL
+  end
+
+  def recreate_metrics_views
+    drop_view("metrics")
+    drop_view("metric_rollups")
```

Comment:
when we drop the view does it automatically drop the triggers?

Reply:
Correct:
Before the migration:
```
vmdb_development=# select trigger_name from information_schema.triggers;
       trigger_name       
--------------------------
 metrics_partition
 metric_rollups_partition
(2 rows)
```

After dropping and recreating the view but without creating the triggers:
```
-- execute("      CREATE VIEW metrics AS SELECT * FROM metrics_base\n")
   -> 0.0023s
-- execute("      ALTER VIEW metrics ALTER COLUMN id SET DEFAULT nextval('metrics_base_id_seq')\n")
   -> 0.0006s
-- execute("      CREATE VIEW metric_rollups AS SELECT * FROM metric_rollups_base\n")
   -> 0.0024s
-- execute("      ALTER VIEW metric_rollups ALTER COLUMN id SET DEFAULT nextval('metric_rollups_base_id_seq')\n")
   -> 0.0006s
== 20210225190611 AddMetricsDiskReadWriteLatency: migrated (0.0150s) ==========
$ rails db
vmdb_development=# select trigger_name from information_schema.triggers;
 trigger_name 
--------------
(0 rows)
```

And then after recreating the views and adding the triggers:
```
-- execute("      CREATE VIEW metrics AS SELECT * FROM metrics_base\n")
   -> 0.0026s
-- execute("      ALTER VIEW metrics ALTER COLUMN id SET DEFAULT nextval('metrics_base_id_seq')\n")
   -> 0.0008s
-- execute("      CREATE VIEW metric_rollups AS SELECT * FROM metric_rollups_base\n")
   -> 0.0024s
-- execute("      ALTER VIEW metric_rollups ALTER COLUMN id SET DEFAULT nextval('metric_rollups_base_id_seq')\n")
   -> 0.0005s
-- add_trigger(:insteadof, :metrics, :metrics_partition)
   -> 0.0017s
-- add_trigger(:insteadof, :metric_rollups, :metric_rollups_partition)
   -> 0.0009s
== 20210225190611 AddMetricsDiskReadWriteLatency: migrated (0.0182s) ==========
$ rails db
vmdb_development=# select trigger_name from information_schema.triggers;
       trigger_name       
--------------------------
 metrics_partition
 metric_rollups_partition
(2 rows)
```